### ISSUE TITLE
Cannot run dotnet core WinForms app on .Net 6.0

### ISSUE THREAD

**Issue by lukos** [dotnet/winforms#6854]
* .NET Core Version: 
6.0.3

* Have you experienced this same bug with .NET Framework?: 
No but not directly comparable.

**Problem description:**
Created a new .Net 6.0 winforms app. Built the published content with:

`dotnet publish EmailTracker.sln --configuration Release --runtime win10-x64 --no-build --self-contained false`

Deployed via CD server onto server and attempted to run the exe. Error message:

![image](https://user-images.githubusercontent.com/581018/158617899-0cc84b0d-c555-4e94-8b1a-80b153e00a25.png)

Click through and download the 6.0.3 desktop runtime. If I run the app again, same error. I tried installing the base .net core runtime too but still the same error. 

Error in event log:

`Message: A fatal error was encountered. The library 'hostpolicy.dll' required to execute the application was not found in 'C:\Program Files\dotnet\'.`

That file does *not* exist in that folder but it does exist further down under dotnet/shared/Microsoft.NETCore.App/6.0.3 (and other versioned folders)

Then it says `Failed to run as a self-contained app`, which is fine because it is not supposed to be self-contained.

**Expected behavior:**
I would expect that it would just work when I run the program with the correct runtime installed. Otherwise a much more specific error message if I've done something too stupid because this is show-stopping stuff for a new framework that worked perfectly until I tried to deploy it. Understand all the publish options, the SDK for building and the various project settings is very confusing so I might have got something wrong there.

**Minimal repro:**
This is just an app with a single form and nothing special to deploy but I expect it won't be easy to recreate otherwise you would see it lots of other places.

Deployment server(s):
Windows Server 2016 Datacenter 64 bit.

**Comment by dreddy-work** [comment]
Unfortunately, this is by design. Any application built against .NET 6.0.3+ will be able to run only on machines with .NET 6.0.3+.  Please see https://github.com/dotnet/winforms/issues/6663 for more details.

**Comment by dreddy-work** [comment]
> Sorry why have you closed this? I built this targeting 6.0.3 and I have installed the runtime for 6.0.3 on the target machines, that is the whole reason I raised this issue, I still get an error.

Apologize, I misread you were trying to run it on 6.0 but not on 6.0.3. Let us try repro this.

**Comment by lukos** [comment]
The 6.0 is just what is available in the project target properties i.e. you can choose core 2.2, 2.2, 3.1 5 or 6

**Comment by Olina-Zhang** [comment]
We cannot repro this issue, after published app with the command "dotnet publish EmailTracker.sln --configuration Release --no-build --self-contained false", on a clean machine, without 6.0.3 installed, a dialog about following downloading pops up. And once we installed it, app can be ran successfully.
![image](https://user-images.githubusercontent.com/26474449/158922113-727264fc-0f6b-4abf-8b79-31cdd50eacfe.png)

**Comment by RussKie** [comment]
I am afraid I can't repo the issue either.

1. created a brand new Windows Forms app
```
dotnet new winforms
```
3. Publish 
```
dotnet publish --configuration Release --runtime win10-x64 --self-contained false
```
3. Copy to a brand new VM that has no .NET runtime or SDK installed
4. Run the app, follow the prompt to download the **desktop** runtime
5. Install the desktop runtime
6. Run the app

@lukos were you following the same steps? If this issue specific to a specific workstation or occurs on different workstations?

**Comment by lukos** [comment]
It fails the same way on a number of deployment targets with the same problem. Did you try on Windows Server 2016? Not sure why it should matter but it might.

The servers also have other dotnet runtimes installed but didn't have desktop 6.0 until I ran the app, used the link and installed the new runtime.

This is a real shame because I have fallen at the first hurdle, and this makes winforms v6.0 unusable for us. There must be a better way to debug why it cannot detect the installed runtime? The complete details from the error logged in the event log are:

```
Description: A .NET application failed.
Application: EmailTracker.exe
Path: C:\Octopus\Applications\EmailTracker\1.0.0.13\EmailTracker.exe
Message: A fatal error was encountered. The library 'hostpolicy.dll' required to execute the application was not found in 'C:\Program Files\dotnet\'.
Failed to run as a self-contained app.
  - The application was run as a self-contained app because 'C:\Octopus\Applications\EmailTracker\1.0.0.13\EmailTracker.runtimeconfig.json' did not specify a framework.
  - If this should be a framework-dependent app, specify the appropriate framework in 'C:\Octopus\Applications\EmailTracker\1.0.0.13\EmailTracker.runtimeconfig.json'.
```

But the runtimeconfig.json DOES include 6.0.3:

```json
{
  "runtimeOptions": {
    "tfm": "net6.0",
    "includedFrameworks": [
      {
        "name": "Microsoft.NETCore.App",
        "version": "6.0.3"
      },
      {
        "name": "Microsoft.WindowsDesktop.App",
        "version": "6.0.3"
      }
    ],
    "configProperties": {
      "System.Reflection.Metadata.MetadataUpdater.IsSupported": false
    }
  }
}
```

Is it possible that the templates in Visual Studio 2022 are not correct? I shouldn't be having to mess around with config files just to get a .net 6 app working out of the box surely?

**Comment by lukos** [comment]
![image](https://user-images.githubusercontent.com/581018/158981871-64204e9a-e6ab-4a1b-addb-81798a68f764.png)

**Comment by RussKie** [comment]
@lukos: yes, these workloads look good. On my VM I only installed windowsdesktop-runtime-6.0.3-win-x64.exe, which installed the following:
![image](https://user-images.githubusercontent.com/4403806/159613066-809120b9-e561-4fae-befd-8d92a0cd1433.png)
![image](https://user-images.githubusercontent.com/4403806/159613213-d959a99c-2b9d-4d63-89b1-bfdc2d794396.png)
![image](https://user-images.githubusercontent.com/4403806/159613550-ae5ce8ab-c5f4-41bb-81c7-8cf1d3dfb3e3.png)

I have hostpolicy.dll located in C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App\6.0.3.

<details>
<summary>WinFormsApp2.runtimeconfig.json</summary>

```json
{
  "runtimeOptions": {
    "tfm": "net6.0",
    "frameworks": [
      {
        "name": "Microsoft.NETCore.App",
        "version": "6.0.3"
      },
      {
        "name": "Microsoft.WindowsDesktop.App",
        "version": "6.0.3"
      }
    ],
    "configProperties": {
      "System.Reflection.Metadata.MetadataUpdater.IsSupported": false
    }
  }
}
```

</details>

<details>
<summary>WinFormsApp2.deps.json</summary>

```json
{
  "runtimeTarget": {
    "name": ".NETCoreApp,Version=v6.0/win10-x64",
    "signature": ""
  },
  "compilationOptions": {},
  "targets": {
    ".NETCoreApp,Version=v6.0": {},
    ".NETCoreApp,Version=v6.0/win10-x64": {
      "WinFormsApp2/1.0.0": {
        "runtime": {
          "WinFormsApp2.dll": {}
        }
      }
    }
  },
  "libraries": {
    "WinFormsApp2/1.0.0": {
      "type": "project",
      "serviceable": false,
      "sha512": ""
    }
  }
}
```

</details>


We've tested Windows Server 2019 and found no issues. We'll see if we can provision a Windows Server 2016 VM to test.

@dsplaisted do you have any thoughts on https://github.com/dotnet/winforms/issues/6854#issuecomment-1072246188?

**Comment by Olina-Zhang** [comment]
Tested it on Windows server 2016, after installed .Net 6.0.3, the published app can be launched successfully:
![image](https://user-images.githubusercontent.com/26474449/159657206-9db4fa2e-3be7-4b08-91bd-22c4676b8927.png)

**Comment by lukos** [comment]
@Olina-Zhang Did you test this with the same build parameters for dotnet publish? I am trying to work out what other difference there could be and why this is not working on our servers.

**Comment by Olina-Zhang** [comment]
@lukos I just removed `--runtime win10-x64` from command `dotnet publish EmailTracker.sln --configuration Release --runtime win10-x64 --no-build --self-contained false` because I can't use it publish app successfully.

**Comment by lukos** [comment]
Thanks @Olina-Zhang I've just found out what wasn't working.

I was passing `--no-build` to `dotnet publish` because we have already run dotnet build and run automated tests so it seemed pointless to build it again as part of publish. I removed that flag and now it works.

I'm not sure if that is strictly a bug but is definitely a gotcha (I have 20 years .Net programming so I'm not a noob) but not sure how this could work differently because in other places (not Winforms) we do it this way and it's fine. Maybe `--self-contained false` should override no-build or something?
